Docs/polishing 2 (#19)
wiktor-obrebski/type-safe-errors: docs/REFERENCE.md

# API Reference

`type-safe-errors` exposes three class-like abstractions: [Ok](#ok), [Err](#err), and [Result](#result).

For an introduction to the type-safe-errors library and its benefits, please refer to the [README](../README.md).  For framework-specific examples, please refer to the [Framework Examples](../examples) directory.

## Ok
An `Ok` object represents a valid result of an action.

### Ok.of(...)

> Creates a new `Ok` result. It is a static function, meaning it can only be called on the imported `Ok` namespace and not on `Ok` result instances.

Reviewer: Here it would be good to also inform about supporting of promises

ex: `Ok.of(Promise.resolve(3))`

Author: but there is not such support. `Ok.of` always expect static value.
`Result.from(() => ...)` can expect async/sync result.

Reviewer: So we can also say that they are forbidden or link it `if you want to use promises use Result.from`

import { Ok } from 'type-safe-errors';

> const okResult = Ok.of({

Reviewer: Może warto też dodać przykład z podaniem typu Ok.of<xxx>

The operation is the results version of [Promise.prototype.catch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch) function.

**Examples:**

```ts
import { Ok } from 'type-safe-errors';

const okResult = Ok.of(5);
// No operation is performed, the original Ok result is returned
const sameOkResult = okResult.mapAnyErr(err => 123); 

```

---

### ok.unsafePromise()

> Map an `Ok` result to a fulfilled [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).  

Reviewer: Swap order of `ok.unsafePromise` and `ok.promise` 
Then it's more clear why it's unsafePromise

## Err
An `Err` object represents an invalid result of an action.

### Err.of(...)

> Creates a new `Err` result. It is a static function, meaning it can only be called on the imported `Err` namespace and not on `Err` result instances.

Reviewer: Same information about `Err.of(promise.resolve())`

Author: there is no such supported syntax that lead to anything meaningful

Reviewer: same argument as for `Ok.of`

### err.map(callback)

No operation is performed for `Err` results. This interface is common for both `Ok` and `Err` results: [result.map(callback)](#resultmapcallback)

**Examples:**

```ts
import { Ok } from 'type-safe-errors';
import { UserNotFoundError } from './errors';

const errResult = Err.of(new UserNotFoundError());
// No operation is performed, the original Err result is returned
const sameErrResult = errResult.map(err => 123); 
```

---

### err.mapErr(ErrorClass, callback)

> Map the `Err` result of specified class to a different result.

Reviewer: ```suggestion
Map the `Err` result of specified class to a different result. If you don't explicitly return value of `Err` type it will be mapped to `Ok`
```

Interface common for both types of results: [result.mapErr(ErrorClass, callback)](#resultmaperrerrorclass-callback)

**Examples:**

```ts
import { Ok } from 'type-safe-errors';
import { UserNotFoundError, Http404Error } from './errors';

const errResult = Err.of(new UserNotFoundError());

// Ok.of('User not found')

> const sameErrResult = errResult.mapErr(UserNotFoundError, err => 'User not found'); 

Reviewer: ```suggestion
const okResultWithString = errResult.mapErr(UserNotFoundError, err => 'User not found'); 
```

// Err result of Http404Error is returned
const httpErrResult = errResult.mapErr(UserNotFoundError, err => Err.of(new Http404Error())); 
```

---

### err.mapAnyErr(callback)

> Map the `Err` result to a different result.

Reviewer: ```suggestion
Map any `Err` result to a different result. If you don't explicitly return value of `Err` type it will be mapped to `Ok`
```

Interface common for both types of results: [result.mapAnyErr(callback)](#resultmapanyerrcallback)

**Examples:**

```ts
import { Ok } from 'type-safe-errors';
import { UserNotFoundError, Http404Error } from './errors';

const errResult = Err.of(new UserNotFoundError());

// Ok.of('User not found')

> const sameErrResult = errResult.mapAnyErr(err => 'User not found'); 

Reviewer: ```suggestion
const okResultWithString = errResult.mapAnyErr(err => 'User not found'); 
```

// Err result of Http404Error is returned
const httpErrResult = errResult.mapAnyErr(err => Err.of(new Http404Error())); 
```
---

### err.unsafePromise()

Map an `Err` result to a rejected [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). This is a common interface for both `Ok` and `Err` results: [result.unsafePromise()](#resultunsafePromise).

**Examples:**

```typescript
import { Err } from 'type-safe-errors';
import { UserNotFoundError } from './errors';

const errResult = Err.of(new UserNotFoundError());
// Promise rejects with value of UserNotFoundError error instance
const rejectedPromise = errResult.unsafePromise();

```

---

## Result

`Result` provides static utility functions to work with multiple results.

### Result.combine([result1, result2, ...])

Combines a list of provided results into a single result. If all provided results are `Ok`, the returned result will be an `Ok` containing an array of values from the provided results: `[Ok<A>, Ok<B>] -> Ok<[A, B]>`.  

> If provided results list have at least one `Err` result, returned result will be `Err` of first `Err` result value found in the array.

Reviewer: Is that still true isn't that sum of error types ?

ex: 
```ts
const ok1Result = Ok.of(5);
const ok2Result = Ok.of(9);
const err1Result = Err.of(new UserNotFoundError());
const err2Result = Err.of(new UserBlocked());
const okSumResult = Result.combine([ok1Result, err1Result , ok2Result, err2Result])
```
` type of: okSumResult === Result<5 | 9, UserNotFoundError | UserBlocked> ?`

Author: its not exactly like that. there is better example to consider:
```ts
const ok1Result = Ok.of(1);
const ok2Result = Ok.of(2);
const maybeErr1Result = Ok.of(5).map(() => Math.random() > 0.5 ? Err.of(new Error1()) : 5);
const maybeErr2Result = Ok.of(6).map(() => Math.random() > 0.5 ? Err.of(new Error2()) : 6);
const okSumResult = Result.combine([ok1Result, maybeErr1Result , ok2Result, maybeErr2Result])
```
`type of: okSumResult === Result<[1, 5, 2, 6], Error1 | Error2>`.

I think you confuse types with execution. The description saying:
> If provided results list have at least one `Err` result, returned result will be `Err` of first `Err` result value found in the array.

In execution it will be first error which will happen. But in types you do not know which error will happen first, so you have potentially all errors available.

Reviewer: maybe if that was not clear for me we can change to your example because it's more descriptive?

Author: I improved combine signature and add more examples.

### result.mapAnyErr(callback)

> Map the `Err` result to another result. 

Reviewer: ```suggestion
Map any `Err` result to another result. 
```
